### PR TITLE
Cex style api

### LINKED ISSUE #14: Client API with a more CEX feel
Here is an example proposal to kick start this [card](https://trello.com/c/nYLga1Kc/59-api-that-is-closer-to-cex-feel), I would appreciate addition/removal/modification/reordering of various items here

* offer an API more aligned with various clients within ccxt (library which supports many popular exhanges), I would choose the [binance ](binance), and, [ftx ](https://github.com/ccxt/ccxt/blob/master/python/ccxt/ftx.py)python clients as example
* work item 1 - offer following basic client methods, for args and response structures I would stick with as much as possible to above clients wherever possible
	* fetch markets
	* fetch balance
	* fetch tickers, fetch ticker
	* fetch orders, fetch order
	* fetch open orders
	* cancel orders, cancel order
	* cancel all orders
* work item 2 - offer following more low level methods from ccxt/binance api
	* `<placeholder>`	
* work item 3 - offer more mango/serum/dex specific methods e.g. support for configuring margin/leverage, borrowing/lending, etc., note: these are not really obvious to me at the moment, I am currently still reading up on the existing client

misc
* should the current client be fully replaced? or would something like MangoClientV2 (or another better name more sense?)
* could we author tests against testnet or devnet? do we have some existing boilerplate code for bootstrapping mango markets on testnet/devnet? 

more future work
* offer integration into ccxt?

----

My motivation for working on this - currently, I have following very basic non-open source python scripts for trading on binance, and I would like to trade similarly on mango/serum

* book profits - book partial profits on positions where I am more than 10% up
* set orders for buying dips - I run this for replacing open orders on a ticker/tickers
* set orders for selling rips - I run this for replacing open orders on a ticker/tickers
* remove all open long orders - helps me close all in one run
* remove all open short orders - helps me close all in one run
* get overview of my balance, open positions and open orders, also how much leverage I am already using, how many not-yet-deployed capital I have

Of course, I would appreciate some more input/inspiration on how others trade programmatically

### PR DESCRIPTION
**Pull request by feierabend654**
Here is an initial draft for what I think is a cex style or rather a simpler API. I am still looking at the tests and there are some todos, but some initial feedback would be welcome. A method called `getPnl` is empty, not sure how to proceed there, so any pointers would be great. 

closes #14

### CODE REVIEW

File: src/simpleclient.ts [hunk 1]
```diff
@@ -0,0 +1,561 @@
...
+    if (!Number.isFinite(price) || price <= 0) {
+      throw new Error(`invalid price ${price}`);
+    }
+
+    const market = this.getMarketForSymbol(symbol);
+
+    const marginAccounts = await this.getMarginAccountsForOwner();
+    // todo: rather choose margin account with largest balance for base or
+    // quote currency based on trade side
+    const marginAccountWithLargestBalance = marginAccounts[0];
```
Comment (mschneider) [inline src/simpleclient.ts:315]: most users will have exactly one margin account (which this client should create on demand imo). users with multiple accounts will have those for the purpose of managing risk or for participation in a trading competition. the users with multiple accounts most likely want to have the control "where their order is placed". what is the scenario you were thinking of solving with this logic?

what do you think about a scoped / more stateful design for the client, which can only be used to work with a single margin account at a time?

File: tests/simpleclient.test.ts [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+  });
+
+  it('test cancel a specific order', async () => {
+    const sc = await SimpleClient.create();
+    await sc.cancelOrders();
+    await sc.placeOrder(marketSymbol, 'buy', 0.0001, 48000);
+    const ordersAfterPlacement = await sc.getOpenOrders(marketSymbol);
+    await sc.cancelOrders(
+      marketSymbol,
+      ordersAfterPlacement[0].orderId.toString(),
```
Comment (mschneider) [inline tests/simpleclient.test.ts:143]: this feels clunky, the result of placeOrder should give you all you need to cancel the same order

File: src/simpleclient.ts [hunk 3]
```diff
@@ -0,0 +1,561 @@
...
+    );
+  }
+
+  /// public
+
+  async placeOrder(
+    symbol: SpotMarketSymbol,
+    side: 'buy' | 'sell',
+    quantity: number,
+    price: number,
```
Comment (mschneider) [inline src/simpleclient.ts:298]: how do we specify market vs limit?

File: src/simpleclient.ts [hunk 4]
```diff
@@ -0,0 +1,561 @@
...
+        marginAccount.publicKey,
+        marginAccountBalances,
+        marketBalances,
+      );
+    });
+  }
+
+  async getPnl() {
+    // todo
+    throw new Error('not implemented');
```
Comment (mschneider) [inline src/simpleclient.ts:482]: We use another service, that is not open source yet, but provides you with a trade history for a specific open orders account. You can see how to use it here: https://github.com/blockworks-foundation/mango-ui/blob/5715e09fee5f5f606a8c86f72f1a622f6307858d/src/utils/useTradeHistory.tsx#L36

And here is code for the PnL (unrealized) calculation https://github.com/blockworks-foundation/mango-ui-v2/blob/5f61a86a615e4c92f53d565f81de1622c615f365/components/MarginInfo.tsx#L11

**Review by mschneider — COMMENTED** [review]
I really like the direction, this looks really good.

The only major issue I see, is the amalgamation of the different margin accounts. Let me know what your pro/cons are here.

File: src/simpleclient.ts [hunk 5]
```diff
@@ -0,0 +1,561 @@
...
+    }
+  }
+
+  async getTradeHistory(symbol: SpotMarketSymbol): Promise<Order[]> {
+    if (!symbol.trim()) {
+      throw new Error(`invalid symbol ${symbol}`);
+    }
+    const market = this.getMarketForSymbol(symbol);
+
+    const fills = await market.loadFills(this.connection, 10_000);
```
[the remark this answers is not in the document]
Comment (feierabend654) [reply src/simpleclient.ts]: done

File: src/simpleclient.ts (hunk 1 above)
Comment (feierabend654) [reply src/simpleclient.ts]: I agree, I changed the code to grab by default largest margin account the user has and implicitly use that. Other than this I create a margin account inline implicitly if none exists

File: tests/simpleclient.test.ts (hunk 2 above)
Comment (feierabend654) [reply tests/simpleclient.test.ts]: fixed and done

File: src/simpleclient.ts (hunk 3 above)
Comment (feierabend654) [reply src/simpleclient.ts]: done

File: src/simpleclient.ts (hunk 4 above)
Comment (feierabend654) [reply src/simpleclient.ts]: done, also added tests

**Comment by mschneider** [comment]
tried building it locally, but got an error:
```
mango-client-ts $ yarn
yarn install v1.22.10
[1/5] 🔍  Validating package.json...
[2/5] 🔍  Resolving packages...
success Already up-to-date.
$ run-s clean build
yarn run v1.22.10
$ rm -rf lib
✨  Done in 0.10s.
yarn run v1.22.10
$ tsc --project ./tsconfig.production.json
src/simpleclient.ts:13:28 - error TS2307: Cannot find module '../lib' or its corresponding type declarations.

13 import { MangoGroup } from '../lib';
                              ~~~~~~~~


Found 1 error.

error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
ERROR: "build" exited with 2.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

**Review by mschneider — APPROVED** [review]
LGTM, please use squash & merge

**Pull request merged**